# 01. Importing Libraries

4.9 Task part 1

In [1]:
# Task Step 1: download customer data set
# Task Step 2: Creating new notebook for task

In [2]:
# Task Step 3a: Importing libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [3]:
# Task Step 3b: Importing customer data set

# 02. Importing data

In [4]:
path = r'/Users/bdigregorio/Desktop/CareerFoundry/' \
r'Data Immersion/Data Immersion 4/DI 4.2/Instacart Basket Analysis'

In [5]:
path

'/Users/bdigregorio/Desktop/CareerFoundry/Data Immersion/Data Immersion 4/DI 4.2/Instacart Basket Analysis'

In [6]:
# Import customers.csv
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)


In [7]:
# Importing newest prods (from 4.8) for task step 6
ords_prods_new = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_new.pkl'))


# 03. Wrangling 

In [8]:
# Task Step 4 

In [9]:
# learning details of customers.csv data set
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [10]:
df_customers.tail()

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095

In [11]:
df_customers.shape

(206209, 10)

In [12]:
df_customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [13]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [14]:
df_customers.isnull().sum()

user_id             0
First Name      11259
Surnam              0
Gender              0
STATE               0
Age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

Observations:
 1. column names can be cleaned up
 2. First_name has 11259 nulls. However, first names cannot be imputed and will not affect analysis. Choose to ignore.

In [15]:
# cleaning column names (for consistency)
df_customers.rename(columns = {'First Name' : 'first_name', 'Surnam' : 'last_name', 'Gender' : 'gender', 'STATE' : 'state', 'Age' : 'age', 'n_dependants' : 'number_of_dependents', 'fam_status' : 'family_status'}, inplace = True)


In [16]:
# check column name changes
df_customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependents family_status  income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

In [17]:
# checking data types
df_customers.dtypes

user_id                  int64
first_name              object
last_name               object
gender                  object
state                   object
age                      int64
date_joined             object
number_of_dependents     int64
family_status           object
income                   int64
dtype: object

In [18]:
# change data type for user_id to 'str'
df_customers['user_id'] = df_customers['user_id'].astype('str')

In [19]:
# check data type change
df_customers['user_id'].dtype

dtype('O')

# 04. Consistency checks

In [20]:
# Task step 5 quality and consistency checks

In [21]:
# Data types and nulls adressed above

In [22]:
# looking for duplicates
df_dups = df_customers[df_customers.duplicated()]

In [23]:
df_dups.head()

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, number_of_dependents, family_status, income]
Index: []

In [24]:
df_customers.duplicated()

0         False
1         False
2         False
3         False
4         False
          ...  
206204    False
206205    False
206206    False
206207    False
206208    False
Length: 206209, dtype: bool

Observation: There are no duplicates

In [25]:
# checking for mixed data types
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

first_name


In [26]:

df_customers['first_name'].dtype

dtype('O')

In [27]:
df_customers['first_name'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 206209 entries, 0 to 206208
Series name: first_name
Non-Null Count   Dtype 
--------------   ----- 
194950 non-null  object
dtypes: object(1)
memory usage: 1.6+ MB


In [28]:
# 'first_name' must have appeared due to missing values. Confirm 'first-name' is 'str'
df_customers['first_name'] = df_customers['first_name'].astype('str')

In [29]:
df_customers['first_name'].dtype

dtype('O')

Confirming first_name blanks will not affect analysis and continue to ignore

In [30]:
ords_prods_new['user_id'].dtype

dtype('float64')

In [31]:
# finding NaNs in ords_prods_new
ords_prods_new['user_id'].isnull().sum()

11

In [32]:
ords_prods_new['user_id'].shape

(32642461,)

In [33]:
ords_prods_new['user_id'].info()

<class 'pandas.core.series.Series'>
Int64Index: 32642461 entries, 0 to 32642460
Series name: user_id
Non-Null Count     Dtype  
--------------     -----  
32642450 non-null  float64
dtypes: float64(1)
memory usage: 498.1 MB


In [34]:
ords_prods_new['user_id'].head()

0    1.0
1    1.0
2    1.0
3    1.0
4    1.0
Name: user_id, dtype: float64

In [35]:
ords_prods_new['user_id'].tail(12)

32642449    205420.0
32642450         NaN
32642451         NaN
32642452         NaN
32642453         NaN
32642454         NaN
32642455         NaN
32642456         NaN
32642457         NaN
32642458         NaN
32642459         NaN
32642460         NaN
Name: user_id, dtype: float64

In [36]:
ords_prods_new.tail(12)

order_id   user_id  order_number  orders_day_of_the_week  \
32642449  3093936.0  205420.0           1.0                     4.0   
32642450        NaN       NaN           NaN                     NaN   
32642451        NaN       NaN           NaN                     NaN   
32642452        NaN       NaN           NaN                     NaN   
32642453        NaN       NaN           NaN                     NaN   
32642454        NaN       NaN           NaN                     NaN   
32642455        NaN       NaN           NaN                     NaN   
32642456        NaN       NaN           NaN                     NaN   
32642457        NaN       NaN           NaN                     NaN   
32642458        NaN       NaN           NaN                     NaN   
32642459        NaN       NaN           NaN                     NaN   
32642460        NaN       NaN           NaN                     NaN   

          order_hour_of_day  days_since_prior_order  product_id  \
32642449               14.0                     NaN     28818.0   
32642450                NaN                     NaN      3630.0   
32642451                NaN                     NaN      3718.0   
32642452                NaN                     NaN      7045.0   
32642453                NaN                     NaN     25383.0   
32642454                NaN                     NaN     27499.0   
32642455                NaN                     NaN     36233.0   
32642456                NaN                     NaN     37703.0   
32642457                NaN                     NaN     43725.0   
32642458                NaN                     NaN     45971.0   
32642459                NaN                     NaN     46625.0   
32642460                NaN                     NaN     49540.0   

          add_to_cart_order  reordered _merge  ...    price_range_loc  \
32642449                8.0        0.0   both  ...                NaN   
32642450                NaN        NaN    NaN  ...  Low-range product   
32642451                NaN        NaN    NaN  ...                NaN   
32642452                NaN        NaN    NaN  ...                NaN   
32642453                NaN        NaN    NaN  ...  Low-range product   
32642454                NaN        NaN    NaN  ...                NaN   
32642455                NaN        NaN    NaN  ...                NaN   
32642456                NaN        NaN    NaN  ...                NaN   
32642457                NaN        NaN    NaN  ...                NaN   
32642458                NaN        NaN    NaN  ...                NaN   
32642459                NaN        NaN    NaN  ...  Low-range product   
32642460                NaN        NaN    NaN  ...                NaN   

             busiest_day    busiest_days     Most orders  max_order  \
32642449      Least busy  Regularly busy     Most orders       17.0   
32642450  Regularly busy  Regularly busy  Average orders        NaN   
32642451  Regularly busy  Regularly busy  Average orders        NaN   
32642452  Regularly busy  Regularly busy  Average orders        NaN   
32642453  Regularly busy  Regularly busy  Average orders        NaN   
32642454  Regularly busy  Regularly busy  Average orders        NaN   
32642455  Regularly busy  Regularly busy  Average orders        NaN   
32642456  Regularly busy  Regularly busy  Average orders        NaN   
32642457  Regularly busy  Regularly busy  Average orders        NaN   
32642458  Regularly busy  Regularly busy  Average orders        NaN   
32642459  Regularly busy  Regularly busy  Average orders        NaN   
32642460  Regularly busy  Regularly busy  Average orders        NaN   

              loyalty_flag avg_spent  spender_type median_order_days  \
32642449  Regular customer  7.684746  High spender              13.0   
32642450               NaN       NaN           NaN               NaN   
32642451               NaN       NaN  High spender               NaN   
32642452               NaN       NaN           NaN   

In [37]:
ords_prods_new['user_id'].dtype

dtype('float64')

In [38]:
df_customers['user_id'].dtype

dtype('O')

In [39]:
# changing df_customers 'user_id' back to 'int64' to facilitate merge
df_customers['user_id'] = df_customers['user_id'].astype('int64')

In [40]:
df_customers['user_id'].dtype

dtype('int64')

In [41]:
df_customers.shape

(206209, 10)

# 05. Merging data

In [42]:
# Task Step 6: Combine your customer data with the rest of your prepared Instacart data


In [43]:
# joining df_customers and ords_prods_new will be on user_id.

In [44]:
ords_prods_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32642461 entries, 0 to 32642460
Data columns (total 26 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                float64 
 1   user_id                 float64 
 2   order_number            float64 
 3   orders_day_of_the_week  float64 
 4   order_hour_of_day       float64 
 5   days_since_prior_order  float64 
 6   product_id              float64 
 7   add_to_cart_order       float64 
 8   reordered               float64 
 9   _merge                  category
 10  Unnamed: 0              float64 
 11  product_name            object  
 12  aisle_id                float64 
 13  department_id           float64 
 14  prices                  float64 
 15  merge_flag              category
 16  price_range_loc         object  
 17  busiest_day             object  
 18  busiest_days            object  
 19  Most orders             object  
 20  max_order               float64 
 21  loyalt

In [45]:
ords_prods_new.rename(columns = {'_merge' : 'ords_priors_combined', 'merge_flag' : 'prods_clened_combined'}, inplace = True)


In [46]:
ords_prods_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32642461 entries, 0 to 32642460
Data columns (total 26 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                float64 
 1   user_id                 float64 
 2   order_number            float64 
 3   orders_day_of_the_week  float64 
 4   order_hour_of_day       float64 
 5   days_since_prior_order  float64 
 6   product_id              float64 
 7   add_to_cart_order       float64 
 8   reordered               float64 
 9   ords_priors_combined    category
 10  Unnamed: 0              float64 
 11  product_name            object  
 12  aisle_id                float64 
 13  department_id           float64 
 14  prices                  float64 
 15  prods_clened_combined   category
 16  price_range_loc         object  
 17  busiest_day             object  
 18  busiest_days            object  
 19  Most orders             object  
 20  max_order               float64 
 21  loyalt

In [47]:
# merging df_customers and ords_prods_new
df_merged = df_customers.merge(ords_prods_new, on = ['user_id'], indicator = True)

In [48]:
# merging checks
df_merged.head()

user_id first_name last_name  gender     state  age date_joined  \
0    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
1    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
2    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
3    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
4    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   

   number_of_dependents family_status  income  ...     busiest_day  \
0                     3       married  165665  ...  Regularly busy   
1                     3       married  165665  ...  Regularly busy   
2                     3       married  165665  ...  Regularly busy   
3                     3       married  165665  ...  Regularly busy   
4                     3       married  165665  ...  Regularly busy   

     busiest_days    Most orders  max_order  loyalty_flag  avg_spent  \
0  Regularly busy  Fewest orders        9.0  New customer   7.988889   
1  Regularly busy    Most orders        9.0  New customer   7.988889   
2  Regularly busy    Most orders        9.0  New customer   7.988889   
3  Regularly busy  Fewest orders        9.0  New customer   7.988889   
4  Regularly busy    Most orders        9.0  New customer   7.988889   

   spender_type  median_order_days     customer_type  _merge  
0   Low spender               20.0  Regular customer    both  
1   Low spender               20.0  Regular customer    both  
2   Low spender               20.0  Regular customer    both  
3  High spender               20.0  Regular customer    both  
4  High spender               20.0  Regular customer    both  

[5 rows x 36 columns]

In [49]:
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32642450 entries, 0 to 32642449
Data columns (total 36 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   user_id                 int64   
 1   first_name              object  
 2   last_name               object  
 3   gender                  object  
 4   state                   object  
 5   age                     int64   
 6   date_joined             object  
 7   number_of_dependents    int64   
 8   family_status           object  
 9   income                  int64   
 10  order_id                float64 
 11  order_number            float64 
 12  orders_day_of_the_week  float64 
 13  order_hour_of_day       float64 
 14  days_since_prior_order  float64 
 15  product_id              float64 
 16  add_to_cart_order       float64 
 17  reordered               float64 
 18  ords_priors_combined    category
 19  Unnamed: 0              float64 
 20  product_name            object  
 21  aisle_

In [50]:
df_merged.shape

(32642450, 36)

In [51]:
# dropping unneeded column _merge
df_merged.drop(columns = ['_merge'])

user_id first_name last_name  gender     state  age date_joined  \
0           26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
1           26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
2           26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
3           26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
4           26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
...           ...        ...       ...     ...       ...  ...         ...   
32642445    80148    Cynthia     Noble  Female  New York   55    4/1/2020   
32642446    80148    Cynthia     Noble  Female  New York   55    4/1/2020   
32642447    80148    Cynthia     Noble  Female  New York   55    4/1/2020   
32642448    80148    Cynthia     Noble  Female  New York   55    4/1/2020   
32642449    80148    Cynthia     Noble  Female  New York   55    4/1/2020   

          number_of_dependents family_status  income  ...    price_range_loc  \
0                            3       married  165665  ...  Mid-range product   
1                            3       married  165665  ...  Mid-range product   
2                            3       married  165665  ...  Mid-range product   
3                            3       married  165665  ...  Mid-range product   
4                            3       married  165665  ...  Mid-range product   
...                        ...           ...     ...  ...                ...   
32642445                     1       married   57095  ...  Low-range product   
32642446                     1       married   57095  ...  Low-range product   
32642447                     1       married   57095  ...  Low-range product   
32642448                     1       married   57095  ...  Low-range product   
32642449                     1       married   57095  ...                NaN   

             busiest_day     busiest_days    Most orders  max_order  \
0         Regularly busy   Regularly busy  Fewest orders        9.0   
1         Regularly busy   Regularly busy    Most orders        9.0   
2         Regularly busy   Regularly busy    Most orders        9.0   
3         Regularly busy   Regularly busy  Fewest orders        9.0   
4         Regularly busy   Regularly busy    Most orders        9.0   
...                  ...              ...            ...        ...   
32642445  Regularly busy  Least busy days    Most orders        5.0   
32642446  Regularly busy  Least busy days    Most orders        5.0   
32642447  Regularly busy  Least busy days  Fewest orders        5.0   
32642448  Regularly busy  Least busy days    Most orders        5.0   
32642449  Regularly busy  Least busy days    Most orders        5.0   

          loyalty_flag  avg_spent  spender_type median_order_days  \
0         New customer   7.988889   Low spender              20.0   
1         New customer   7.988889   Low spender              20.0   
2         New customer   7.988889   Low spender              20.0   
3         New customer   7.988889  High spender              20.0   
4         New customer   7.988889  High spender              20.0   
...                ...        ...           ...               ...   
32642445  New customer   3.886667   Low spender              12.0   
32642446  New customer   3.886667   Low spender              12.0   
32642447  New customer   3.886667   Low spender              12.0   
32642448  New customer   3.886667   Low spender              12.0   
32642449  New customer   3.886667   Low spender              12.0   

             customer_type  
0         Regular customer  
1         Regular customer  
2         Regular customer  
3         Regular customer  
4         Regular customer  
...                    ...  
32642445  Regular customer  
32642446  Regular customer  
32642447  Regular customer  
32642448  Regular customer  
32642449  Regular customer  

[32642450 rows x 35 columns]

# 05. Exporting
df-merged as "Orders_products_all"

In [52]:
# Task Step 7,8 and 9

In [53]:
df_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'Orders_products_all.pkl'))
